pytorch_lightning/plugins/plugins_registry.py

```diff
@@ -0,0 +1,94 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections import UserDict
+from typing import Callable, List, Optional
+
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+
+class _TrainingTypePluginsRegistry(UserDict):
+    """
+    This class is a Registry that stores information about the Training Type Plugins.
+
+    The Plugins are mapped to strings. These strings are names that idenitify
+    a plugin, eg., "deepspeed". It also returns Optional description and
+    parameters to initialize the Plugin, which were defined durng the
+    registeration.
+
+    The motivation for having a TrainingTypePluginRegistry is to make it convenient
+    for the Users to try different Plugins by passing just strings
+    to the plugins flag to the Trainer.
+
+    Example::
+
+        @TrainingTypePluginsRegistry.register("lightning", description="Super fast", a=1, b=True)
+        class LightningPlugin:
+            def __init__(self, a, b):
+                ...
+
+    """
+
```

Comment:
this opens up a new surface area for backwards compatibility, so we should have clear ways to mark registry entries as deprecated

Reply:
> so we should have clear ways to mark registry entries as deprecated

@ananthsub What do you mean by this?

Reply:
@kaushikb11 we'll need some annotation as to whether a specific plugin configuration is deprecated. For example, if we want to remove a plugin or a specific configuration, we should give users enough heads up and apply our normal rules for deprecating artifacts.

That means we'll need more metadata associated with the registry entry, such as the description, but also the status of the entry. e.g. if its active or deprecated. this way, when we're reading the entries or instantiating items from the registry, we can warn users if they're using a deprecated plugin

Reply:
@ananthsub Not sure about this. I think we can only list the active ones here and still handle the deprecation within the Trainer/AcceleratorConnector as we do now.

Reply:
for deprecating configurations why not:

```python
        @TrainingTypePluginsRegistry.register("myplugin", description="Super fast", a=1, b=True, deprecate=True, deprecate_version=1.5)
```